### PR TITLE
Add bytearray wrapper type

### PR DESCRIPTION
**Pull request by noahbkim**
I tried to follow the precedent of `nanobind::bytes` as closely as seemed reasonable. RFC on the documentation and signature of `resize`. Thanks @oremanj for feedback on the first iteration.

### CODE REVIEW

File: include/nanobind/nb_types.h [hunk 1]
```diff
@@ -443,6 +443,33 @@ class bytes : public object {
...
+class bytearray : public object {
+    NB_OBJECT(bytearray, object, "bytearray", PyByteArray_Check)
+
+    bytearray()
+        : object(PyObject_CallNoArgs((PyObject *)&PyByteArray_Type), detail::steal_t{}) { }
+
+    explicit bytearray(handle h)
+        : object(detail::bytearray_from_obj(h.ptr()), detail::steal_t{}) { }
+
+    explicit bytearray(const char *s, size_t n)
```
Comment (oremanj) [inline include/nanobind/nb_types.h:455]: This should take `const void *`.

File: docs/api_core.rst [hunk 2]
```diff
@@ -1052,6 +1052,40 @@ Wrapper classes
...
+
+      Return the size in bytes.
+
+   .. cpp:function:: const void * data() const
+
+      Convert a Python ``bytearray`` object into a byte buffer of length :cpp:func:`bytearray::size()` bytes.
+
+   .. cpp:function:: void resize(size_t n)
+
+      Resize the internal buffer of a Python ``bytearray`` object to ``n``.
```
Comment (oremanj) [inline docs/api_core.rst:1086]: The note about uninitialized data should be included here as well.

File: include/nanobind/nb_types.h [hunk 3]
```diff
@@ -443,6 +443,33 @@ class bytes : public object {
...
+        : object(detail::bytearray_from_cstr_and_size(s, n), detail::steal_t{}) { }
+
+    const char *c_str() const { return PyByteArray_AsString(m_ptr); }
+
+    const void *data() const { return (const void *) PyByteArray_AsString(m_ptr); }
+
+    size_t size() const { return (size_t) PyByteArray_Size(m_ptr); }
+
+    /// Any space added by this method, which calls `PyByteArray_Resize`, will
+    /// not be initialized and may contain random data.
```
Comment (oremanj) [inline include/nanobind/nb_types.h:465]: This comment can probably be removed if it's mentioned in the actual docs. If you include it, nanobind style would have it as a multiline `/* */` comment.

File: include/nanobind/nb_types.h [hunk 4]
```diff
@@ -443,6 +443,33 @@ class bytes : public object {
     size_t size() const { return (size_t) PyBytes_Size(m_ptr); }
 };
 
+class bytearray : public object {
+    NB_OBJECT(bytearray, object, "bytearray", PyByteArray_Check)
+
+    bytearray()
+        : object(PyObject_CallNoArgs((PyObject *)&PyByteArray_Type), detail::steal_t{}) { }
```
Comment (oremanj) [inline include/nanobind/nb_types.h:450]: It appears from CI that `PyObject_CallNoArgs` doesn't exist in 3.8. You might need an #ifdef here. The alternative would be `PyObject_CallObject` with a second argument of NULL. `CallNoArgs` is documented as faster on the versions where it exists, so we should probably keep it for 3.9 and later.

Comment (noahbkim) [reply include/nanobind/nb_types.h:451]: I'll mark this as done when CI completes, though I also tested it manually.

**Review by oremanj — APPROVED** [review]
LGTM for first pass, over to @wjakob.

**Comment by wjakob** [comment]
(Seems that there is a merge conflict.)

File: docs/changelog.rst [hunk 5]
```diff
@@ -28,6 +28,8 @@ Version 2.0.1 (TBA)
   :cpp:struct:`nb::new_ <new_>`, passed in the same way they would be with
   :cpp:struct:`nb::init <init>`. (issue `#668
   <https://github.com/wjakob/nanobind/issues/668>`__)
+  
+* Added a wrapper type :cpp:class:`bytearray`.
```
Comment (oremanj) [inline docs/changelog.rst:32]: Add a link to this PR.
```
(PR `#654 <https://github.com/wjakob/nanobind/pull/654>`__)
```

**Pull request merged**